pydarn/plotting/fan.py

```diff
@@ -0,0 +1,248 @@
+        scan_no = 0
+        beam_scan = np.zeros((len(dmap_data)))
+        for beam in range(len(dmap_data)):
+            if abs(scan_mark[beam]) == 1:
+                no_scans += 1
+                beam_scan[beam] = no_scans
+            if scan_mark[beam] == 0:
+                beam_scan[beam] = no_scans
+        no_scans += 1
+
+        # Locate scan in loaded data
+        plot_beams = np.where(beam_scan == scan_index)
+
+        # Files holding radar beam/gate locations
+        beam_lats = base_path+'/radar_fov_files/' + \
+            dmap_data[0]['stid'].astype('str').zfill(3)+'_lats.txt'
+        beam_lons = base_path+'/radar_fov_files/' + \
+            dmap_data[0]['stid'].astype('str').zfill(3)+'_lons.txt'
+
+        # Initialise arrays holding coordinates and MLT's
+        beam_corners_lats = np.loadtxt(beam_lats)
+        beam_corners_lons = np.loadtxt(beam_lons)
+
+        fan_shape = beam_corners_lons.shape
+
+        beam_corners_aacgm_lons = np.zeros((fan_shape[0], fan_shape[1]))
+        beam_corners_aacgm_lats = np.zeros((fan_shape[0], fan_shape[1]))
+        beam_corners_mlts = np.zeros((fan_shape[0], fan_shape[1]))
+
+        # Time for coordinate conversion
+        dtime = dt.datetime(dmap_data[plot_beams[0][0]]['time.yr'],
+                            dmap_data[plot_beams[0][0]]['time.mo'], dmap_data[plot_beams[0][0]]['time.dy'],
+                            dmap_data[plot_beams[0][0]]['time.hr'], dmap_data[plot_beams[0][0]]['time.mt'],
+                            dmap_data[plot_beams[0][0]]['time.sc'])
+        for x in range(fan_shape[0]):
+            for y in range(fan_shape[1]):
+                # Convert to AACGM
+                geomag = np.array(aacgmv2.
+                                  get_aacgm_coord(beam_corners_lats[x, y],
+                                                  beam_corners_lons[x, y],
+                                                  250, dtime))
+
+                beam_corners_aacgm_lats[x, y] = geomag[0]
+                beam_corners_aacgm_lons[x, y] = geomag[1]
+
+        # Work out shift due in MLT
+        mltshift = beam_corners_lons[x, y] - \
+            (aacgmv2.convert_mlt(beam_corners_lons[x, y], dtime) * 15)
+        beam_corners_mlts = beam_corners_aacgm_lons - mltshift
+
+        # Hold the beam positions
+        thetas = np.radians(beam_corners_mlts)
+        rs = beam_corners_aacgm_lats
+
+        # Get range-gate data and groundscatter array for given scan
+        scan = np.zeros((fan_shape[0] - 1, fan_shape[1]-1))
+        grndsct = np.zeros((fan_shape[0] - 1, fan_shape[1]-1))
+        iterat = 0
+        for i in np.nditer(plot_beams):
+            try:
+                slist = dmap_data[i.astype(int)]['slist']
+                scan[slist, iterat] = dmap_data[i.astype(int)][parameter]
+                grndsct[slist, iterat] = dmap_data[i.astype(int)]['gflg']
+                iterat += 1
+            except KeyError:
```

Comment:
Why is this `KeyError` here? Please leave a comment :)

Reply:
Er... I was mostly copying the template from your code :P I don't actually know what it does, if it does something could you explain?

Reply:
I think I had this for 'slist' haha well I guess I need to update that too